File: sdks/typescript/src/introspection-client.ts
/** RFC 7662 §2.2 — result of a token introspection request. */
export interface IntrospectionResult {
  /** Whether the token is currently active. */
  active: boolean;
  /** Subject identifier (when active). */
  sub?: string;
  /** Expiration time as Unix seconds (when active). */
  exp?: number;
  /** Issued-at time as Unix seconds (when active). */
  iat?: number;
  /** Issuer identifier (when active). */
  iss?: string;
  /** Intended audience (when active). */
  aud?: string | string[];
  /** Space-separated scope string (when active and present). */
  scope?: string;
  /** OAuth client that requested the token (when active and present). */
  client_id?: string;
  /** All non-standard claims. */
  [key: string]: unknown;
}

/** Configuration for {@link IntrospectionClient}. */
export interface IntrospectionClientConfig {
  /** RFC 7662 introspection endpoint URL. */
  introspectionEndpoint: string;
  /** Client ID used for HTTP Basic authentication. */
  clientId: string;
  /** Client secret used for HTTP Basic authentication. */
  clientSecret: string;
  /** Timeout for outbound HTTP calls in milliseconds. Default: 10 000. */
  httpTimeout?: number;
}

/**
 * Low-level RFC 7662 token introspection client.
 *
 * Results are never cached — per RFC 7662 §2.1, token state can change
 * at any time. Full error taxonomy will be added in §3.
 */
export class IntrospectionClient {
  private readonly endpoint: string;
  private readonly clientId: string;
  private readonly clientSecret: string;
  readonly httpTimeout: number;

  constructor(config: IntrospectionClientConfig) {
    this.endpoint = config.introspectionEndpoint;
    this.clientId = config.clientId;
    this.clientSecret = config.clientSecret;
    this.httpTimeout = config.httpTimeout ?? 10_000;
  }

  /** Introspect a token. Never cached per RFC 7662 §2.1. */
  async introspect(token: string): Promise<IntrospectionResult> {
    const credentials = btoa(`${this.clientId}:${this.clientSecret}`);
    const resp = await fetch(this.endpoint, {
      method: "POST",
      headers: {
        Authorization: `Basic ${credentials}`,
        "Content-Type": "application/x-www-form-urlencoded",
      },
      body: new URLSearchParams({ token }),
      signal: AbortSignal.timeout(this.httpTimeout),
    });
    if (!resp.ok) {
      throw new Error(`Introspection endpoint returned HTTP ${resp.status}`);
    }
    return resp.json() as Promise<IntrospectionResult>;
  }
}

File: sdks/kotlin/hearth-core/src/main/kotlin/io/hearth/sdk/IntrospectionClient.kt
package io.hearth.sdk

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.boolean
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.longOrNull
import okhttp3.FormBody
import okhttp3.OkHttpClient
import okhttp3.Request
import java.util.Base64

/**
 * RFC 7662 token introspection client.
 *
 * Results are **never cached** — per RFC 7662 §2.1, token state can change at any time.
 * Credentials are sent via HTTP Basic (client_id:client_secret) as per RFC 7662 §2.1.
 *
 * Tokens and secrets never appear in thrown error messages per sdk-spec §11.
 */
class IntrospectionClient(
    private val introspectionEndpoint: String,
    private val clientId: String,
    private val clientSecret: String,
    private val httpClient: OkHttpClient,
) {
    /**
     * Introspects [token] at the introspection endpoint.
     *
     * Results are never cached. Returns an [IntrospectionResult] with [IntrospectionResult.active]
     * set to false when the token is unknown or invalid.
     *
     * @throws IntrospectionError when the endpoint is unreachable or returns a non-2xx response.
     */
    suspend fun introspect(token: String): IntrospectionResult = withContext(Dispatchers.IO) {
        val credentials = Base64.getEncoder()
            .encodeToString("$clientId:$clientSecret".toByteArray())

        val formBody = FormBody.Builder()
            .add("token", token)
            .build()

        val request = Request.Builder()
            .url(introspectionEndpoint)
            .post(formBody)
            .header("Authorization", "Basic $credentials")
            .build()

        val response = try {
            httpClient.newCall(request).execute()
        } catch (e: Exception) {
            throw IntrospectionError("Introspection endpoint unreachable", e)
        }

        response.use { resp ->
            if (!resp.isSuccessful) {
                throw IntrospectionError("Introspection endpoint returned HTTP ${resp.code}")
            }
            val bodyStr = resp.body?.string()
                ?: throw IntrospectionError("Introspection endpoint returned empty body")

            parseIntrospectionResult(bodyStr)
        }
    }

    private fun parseIntrospectionResult(bodyStr: String): IntrospectionResult {
        val jsonObj: JsonObject = try {
            JSON.parseToJsonElement(bodyStr).jsonObject
        } catch (e: Exception) {
            throw IntrospectionError("Introspection endpoint returned invalid JSON", e)
        }

        // Known standard RFC 7662 fields — everything else goes into `extra`
        val knownKeys = setOf("active", "sub", "exp", "iat", "iss", "aud", "scope", "client_id")
        val extra = jsonObj.entries
            .filter { it.key !in knownKeys }
            .associate { it.key to it.value }

        return IntrospectionResult(
            active = jsonObj["active"]?.jsonPrimitive?.boolean ?: false,
            sub = jsonObj["sub"]?.jsonPrimitive?.contentOrNull,
            exp = jsonObj["exp"]?.jsonPrimitive?.longOrNull,
            iat = jsonObj["iat"]?.jsonPrimitive?.longOrNull,
            iss = jsonObj["iss"]?.jsonPrimitive?.contentOrNull,
            aud = jsonObj["aud"],
            scope = jsonObj["scope"]?.jsonPrimitive?.contentOrNull,
            clientId = jsonObj["client_id"]?.jsonPrimitive?.contentOrNull,
            extra = extra,
        )
    }
}
